test/test_foreach.py

```diff
@@ -900,6 +900,76 @@ def test_outplace_with_invalid_grads(self, device, dtype, op):
         self.assertIsNotNone(sample.input[0].grad)
         self.assertIsNone(sample.input[1].grad)
 
+    @unittest.skipIf(not (torch.cuda.is_available() and torch.cuda.device_count() > 1), "requires multiple GPUs")
+    def test_tensros_grouping(self):
+
+        from collections import defaultdict
+        import random
+
+        # Moved def of
+        # https://github.com/pytorch/pytorch/blob/843ead134ca6f168df98912efaeea72a0b380909/torch/utils/_foreach_utils.py#L21
+        @torch.no_grad()
```

Comment:
This tests well against the old definition of the grouping logic, but I wonder if it is better to test against invariants instead. (What if the old implementation was missing something?) For example, I would expect:
- For each group in the return values, the devices and dtypes are actually the same.
- Nones are handled.
- When with_indices, you should be able to map the indices to the original tensors (like the tensor with a certain index should be the same).
- The total number of elements in the tensorlists should be the same before and after grouping

What are your thoughts?

Reply:
ah so you meant something solider rather than one that guarantees the parity of the existing one and the new cpp based one

Reply:
Yes.

Reply:
> * For each group in the return values, the devices and dtypes are actually the same.

Isn't this either device_check or https://github.com/pytorch/pytorch/blob/1d5577b601cd88020147e9f5fdfc7bf2b2acb0ae/aten/src/ATen/native/ForeachUtils.h#L68 would guarantee?

> * Nones are handled.

Wouldn't this be verified by making sure the parity of the old one and the new one?

> * The total number of elements in the tensorlists should be the same before and after grouping

Could you elaborate? Ideally with a snippet

Reply:
I do have confidence your current code is correct and does the right things, but the point of the unit test would be to ensure that future changes will not break the invariants we're expecting. I am envisioning a test that looks like:
```
        num_tensors_per_list = 10
        num_devices = torch.cuda.device_count()
        dtypes = (torch.float16, torch.float32, torch.bfloat16)
        list1 = [
                torch.tensor(
                    i,
                    device=torch.device("cuda", random.randint(0, num_devices - 1)),
                    dtype=dtypes[random.randint(0, 2)],
                ) for i in range(num_tensors_per_list)
        ]
        list2 = [None for _ in list1]
        list3 = [torch.rand_like(t) for t in list1]
        nested_tensorlists = [list1, list2, list3]

        grouped_tensors = torch.utils._foreach_utils._group_tensors_by_device_and_dtype(nested_tensorlists, with_indices=True)
        num_tensors_seen = 0
        for (device, dtype), ([l1, l2, l3], indices) in grouped_tensors:
            for t in itertools.chain(l1, l2, l3):
                self.assertEquals(t.device, device)
                self.assertEquals(t.dtype, dtype)
                num_tensors_seen += 1
            self.assertEquals(len(l2), 0)
            for i, index in enumerate(indices):
                 self.assertEquals(l1[i], list1[index])
                 self.assertEquals(l3[i], list3[index])
       self.assertEquals(num_tensors_seen, 2 * num_tensors_per_list)
```
Disclaimer: I just typed this up in the comment so no promises on syntax

Reply:
I'm not quite sure if we do want to corroborate all the tensors of the same index have the same device and the same dtype in general. For optimizers, I think we should though.

Reply:
I think that's the current assumptions we make and would be good to enforce until another use case comes up.

Reply:
Then shouldn't it be an error?

Reply:
I'm okay with that too. Either way, it would be good to capture that intent in code somewhere

Reply:
To clarify, adding the unit test to ensure the invariants would still be better than holding my previous implementation as truth. And I see you've added the TORCH_CHECK, but it is still important to test that the code groups correctly so that future changes to the code won't silently regress the current behavior.